scheduler/src/ratelimit/token_bucket_filter.clj

```diff
@@ -0,0 +1,88 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns ratelimit.token-bucket-filter)
+
+(defn create-tbf
+  "Make a token bucket filter (tbf), with a given refill rate (units per timestep), a maximum token capacity, and give it the
+  current timestamp.
+
+  This token bucket filter is a bit unusual, instead of earning tokens that you can later spend, we allow the current
+  tokens to be negative. Thus, you can (in one request) spend an unbounded number of tokens. However, you will be
+  in debt until your token balance is at least 0. This design allows the tokens in a single request to exceed the max-tokens
+  in the tbf. (If we didn't do it this way, max-tokens is an implicit maximum limit on request size.) This way
+  we can independently pick the burst size in the tbf and the maximum request size.
+
+  This library does not handle blocking or other operations. Typical workflows are to:
+
+  ** Admission control **
+  This is good when you don't know how big a request will be until afterwards.
+
+  Use in-debt? as part of request admission control, rejecting a request if it returns true. When a request completes (or is aborted),
+  take the tokens that it used, affecting the next request.
+
+  ** Delaying requests **
+  If you want to delay requests instead of rejecting. Take the tokens the request needs, pause for the returned pause time.
+  (E.g., schedule the message to be sent at that particular pause time.
+  "
+  ([^double token-rate ^long max-tokens ^long current-time ^long current-tokens]
+   {:pre [(> token-rate 0)
+          (>= max-tokens 0)
+          (>= current-time 0)]}
+   {:token-rate token-rate :max-tokens max-tokens :last-update current-time :current-tokens current-tokens})
+  ([^double token-rate ^long max-tokens ^long current-time]
+   (create-tbf token-rate max-tokens current-time 0)))
+
+(defn- add-tokens-helper
```

Comment:
Nitpick: rename to `update-tbf` to better reflect what the function it is helping.